File: app/src/main/java/com/github/overpass/gather/model/repo/confirm/ConfirmEmailRepo.java
package com.github.overpass.gather.model.repo.confirm;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import com.github.overpass.gather.auth.register.confirm.ConfirmEmailStatus;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;

public class ConfirmEmailRepo {

    private final FirebaseAuth firebaseAuth;

    public ConfirmEmailRepo(FirebaseAuth firebaseAuth) {
        this.firebaseAuth = firebaseAuth;
    }

    public LiveData<ConfirmEmailStatus> confirmEmail() {
        MutableLiveData<ConfirmEmailStatus> confirmEmailStatus = new MutableLiveData<>();
        FirebaseUser firebaseUser = firebaseAuth.getCurrentUser();
        if (firebaseUser == null) {
            Throwable throwable = new Throwable("Something went wrong");
            confirmEmailStatus.setValue(new ConfirmEmailStatus.Error(throwable));
            return confirmEmailStatus;
        }
        firebaseUser.reload()
                .addOnSuccessListener(result -> {
                    confirmEmailStatus.setValue(isEmailVerified());
                })
                .addOnFailureListener(e -> {
                    confirmEmailStatus.setValue(new ConfirmEmailStatus.Error(e));
                });
        return confirmEmailStatus;
    }

    private ConfirmEmailStatus isEmailVerified() {
        FirebaseUser user = firebaseAuth.getCurrentUser();
        if (user != null && user.isEmailVerified()) {
            return new ConfirmEmailStatus.Success();
        } else {
            Throwable throwable = new Throwable("Sorry, your email hasn't been verified");
            return new ConfirmEmailStatus.Error(throwable);
        }
    }
}

File: app/src/main/java/com/github/overpass/gather/model/repo/confirm/ConfirmEmailRepo.kt
package com.github.overpass.gather.model.repo.confirm

import com.github.overpass.gather.model.commons.exception.NotAuthorized
import com.google.android.gms.tasks.Task
import com.google.android.gms.tasks.Tasks
import com.google.firebase.auth.FirebaseAuth

class ConfirmEmailRepo(private val firebaseAuth: FirebaseAuth) {

    fun isEmailVerified(): Boolean {
        return firebaseAuth.currentUser
                ?.isEmailVerified
                ?: false
    }

    fun confirmEmail(): Task<Void> {
        return firebaseAuth.currentUser
                ?.reload()
                ?: Tasks.forException(NotAuthorized())
    }
}
